### PR TITLE
 New ``mode`` argument replacing ``direction``

### PR DESCRIPTION
**Pull request by thet**
Continuation of #29 as an long overdue attempt to get this ready for merging.

Difference to #29 is that old names are not deprecated but simply alternative names.

Reason behind is that all Plone devs are used to the old, confusing names and no unnecessary migration work should be necessary.

In that sense I'm aksing myself if the ``direction`` argument should also not be deprecated and be used along with ``mode`` - but that could cause confusion on it's own.

Open for discussion, if this makes sense or not.

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -19,6 +19,8 @@ Bug fixes:
 3.0.2 (2018-09-28)
 ------------------
 
+Changes:
+
```
Comment (thet) [inline CHANGES.rst]: Oh, forgot to remove that. I'm not doing a new commit to save the planet from unnecessary computing power but directly fix it before an eventual merge.

File: plone/scale/scale.py [hunk 2]
```diff
@@ -129,31 +129,33 @@ def _scale_thumbnail(image, width=None, height=None):
...
-    `down`
+    `contain`
+        Alternative spellings: `scale-crop-to-fit`, `down`
         scaling starts by scaling the largest dimension to the required
         size and crops the other dimension if needed.
 
-    `thumbnail`
-        scales to the requested dimensions without cropping. Theresulting
+    `cover`
+        Alternative spellings: `scale-crop-to-fill`, `up`
```
Comment (thet) [inline plone/scale/scale.py]: Add a dot at the end of this list.

File: plone/scale/scale.py [hunk 3]
```diff
@@ -129,31 +129,33 @@ def _scale_thumbnail(image, width=None, height=None):
...
+        scaling scales the smallest dimension up to the required size
+        and crops the other dimension if needed.
+
+    `scale`
+        Alternative spellings: `keep`, `thumbnail`
```
Comment (thet) [inline plone/scale/scale.py]: Same her.

File: plone/scale/scale.py [hunk 4]
```diff
@@ -129,31 +129,33 @@ def _scale_thumbnail(image, width=None, height=None):
...
+    the CSS background-size values
+    (see https://developer.mozilla.org/en-US/docs/Web/CSS/background-size):
 
-    `up`
-        scaling scales the smallest dimension up to the required size
-        and crops the other dimension if needed.
-
-    `down`
+    `contain`
+        Alternative spellings: `scale-crop-to-fit`, `down`
```
Comment (thet) [inline plone/scale/scale.py]: Same like below.

File: plone/scale/tests/test_scale.py [hunk 5]
```diff
@@ -219,6 +217,58 @@ def testResultBuffer(self):
...
+        """Test alternative and deprecated mode spellings and the old
+        ``direction`` arguments instead of ``mode``.
+        """
+        
+        # scale-crop-to-fit
+        img = PIL.Image.new('RGB', (20, 20), (0, 0, 0))
+        img_scaled = scalePILImage(img, 10, 5, direction='scale-crop-to-fit')
+        self.assertEqual(img_scaled.size, (10, 5))
+        # down
+        img = PIL.Image.new('RGB', (20, 20), (0, 0, 0))
```
Comment (thet) [inline plone/scale/tests/test_scale.py]: By the way: I noticed that the ``scalePILImage`` method scales the image in place. If possible, this should only return a new image while keeping the old one as-is.
Not part of this PR though.

**Comment by thet** [comment]
@jenkins-plone-org please run jobs

File: plone/scale/scale.py [hunk 6]
```diff
@@ -167,12 +169,23 @@ def scalePILImage(image, width=None, height=None, direction='down'):
...
+    del direction
+
+    if mode in ("scale-crop-to-fit", "down"):
+        mode = "contain"
+    elif mode in ("scale-crop-to-fill", "up"):
+        mode = "cover"
+    elif mode in ("keep", "thumbnail"):
+        mode = "scale"
+    try:
+        assert(mode in ('contain', 'cover', 'scale'))
```
Comment (fschulze) [inline plone/scale/scale.py]: This won't work with ``python -O``, is that intended? If not it should be a simple ``if``.

File: plone/scale/tests/test_scale.py [hunk 7]
```diff
@@ -219,6 +216,34 @@ def testResultBuffer(self):
         self.assertEqual(result, img2)      # the return value _is_ the buffer
         self.assertEqual(result.getvalue(), img1)   # but with the same value
 
+    def testAlternativeSpellings(self):
+        """Test alternative and deprecated mode spellings and the old
+        ``direction`` arguments instead of ``mode``.
+        """
+        __import__('pdb').set_trace()
```
Comment (fschulze) [inline plone/scale/tests/test_scale.py]: Stray pdb.set_trace()

File: plone/scale/tests/test_scale.py [hunk 8]
```diff
@@ -219,6 +216,34 @@ def testResultBuffer(self):
...
+        img = Image.new('RGB', (20, 20), (0, 0, 0))
```
Comment (fschulze) [inline plone/scale/tests/test_scale.py]: Missing ``PIL.`` before ``Image``

Comment (fschulze) [reply plone/scale/tests/test_scale.py]: Actually this test doesn't work as is.

**Comment by fschulze** [comment]
Here are fixes for the tests:
```diff
diff --git a/plone/scale/tests/test_scale.py b/plone/scale/tests/test_scale.py
index 31e01c5..e355e1d 100644
--- a/plone/scale/tests/test_scale.py
+++ b/plone/scale/tests/test_scale.py
@@ -220,8 +220,9 @@ class ScalingTests(TestCase):
         """Test alternative and deprecated mode spellings and the old
         ``direction`` arguments instead of ``mode``.
         """
-        __import__('pdb').set_trace()
-        img = Image.new('RGB', (20, 20), (0, 0, 0))
+        src = PIL.Image.new('RGB', (20, 20), (0, 0, 0))
+        img = StringIO()
+        src.save(img, "PNG")
         img_scaled = scaleImage(img, 10, 5, direction='down')
 
         img_scaled = scaleImage(img, 10, 5, direction='scale-crop-to-fit')
@@ -235,10 +236,13 @@ class ScalingTests(TestCase):
         img_scaled = scaleImage(img, 10, 10, direction='thumbnail')
 
     def testDeprecations(self):
+        import plone.scale.scale
+        # clear warnings registry, so the test actually sees the warning
+        plone.scale.scale.__warningregistry__.clear()
         with warnings.catch_warnings(record=True) as w:
             warnings.simplefilter("always")
             scaleImage(PNG, 16, 16, direction="keep")
-            self.assertEqual(len(w), 2)
+            self.assertEqual(len(w), 1)
             self.assertIs(w[0].category, DeprecationWarning)
             self.assertIn(
                 "the 'direction' option is deprecated",
```
I wasn't sure if it is ok to just push them to your branch.

File: plone/scale/scale.py (hunk 6 above)
Comment (thet) [reply plone/scale/scale.py]: Ups, didn't know that....

File: plone/scale/tests/test_scale.py (hunk 7 above)
Comment (thet) [reply plone/scale/tests/test_scale.py]: Somehow my test fixed didn't make it to github. fixing...

File: plone/scale/scale.py (hunk 6 above)
Comment (thet) [reply plone/scale/scale.py]: fixed.

File: plone/scale/tests/test_scale.py (hunk 8 above)
Comment (thet) [reply plone/scale/tests/test_scale.py]: fixed

**Comment by thet** [comment]
@jenkins-plone-org please run jobs

**Comment by thet** [comment]
@jenkins-plone-org please run jobs

File: plone/scale/scale.py [hunk 9]
```diff
@@ -129,31 +129,33 @@ def _scale_thumbnail(image, width=None, height=None):
...
 
-    `up`
-        scaling scales the smallest dimension up to the required size
-        and crops the other dimension if needed.
-
-    `down`
+    `contain`
+        Alternative spellings: `scale-crop-to-fit`, `down`.
         scaling starts by scaling the largest dimension to the required
         size and crops the other dimension if needed.
```
Comment (mauritsvanrees) [inline plone/scale/scale.py]: I recently did some tests because I could not make sense of the up/down difference. So this PR is helpful, and the documentation fits better with what actually happens. But my conclusions were slightly different. I could be wrong.
For `down` my conclusion was:

"scaling starts by scaling the *smallest* dimension to the required size and crops the other dimension if needed."

File: plone/scale/scale.py [hunk 10]
```diff
@@ -129,31 +129,33 @@ def _scale_thumbnail(image, width=None, height=None):
...
 
-    `thumbnail`
-        scales to the requested dimensions without cropping. Theresulting
+    `cover`
+        Alternative spellings: `scale-crop-to-fill`, `up`.
+        scaling scales the smallest dimension up to the required size
+        and crops the other dimension if needed.
```
Comment (mauritsvanrees) [inline plone/scale/scale.py]: For `up` my conclusion was:

"scaling scales the *largest* dimension up to the required size and crops the other dimension if needed."

Comment (fschulze) [reply plone/scale/scale.py]: @mauritsvanrees I think you might be right. Could you take a look at #37 where I put the size calculations into a separate function, it might help to think this through better. With ``up`` and ``down`` you never get letter boxing but if the destination size doesn't fit it crops, the destinations sizes are never exceeded. With ``scale`` you will never get cropping, but the image might not fill the target dimensions and thus you get letter boxing.

File: plone/scale/tests/test_scale.py (hunk 5 above)
Comment (thet) [reply plone/scale/tests/test_scale.py]: See: https://github.com/plone/plone.scale/issues/38

**Comment by thet** [comment]
Replaced by: https://github.com/plone/plone.scale/pull/38

File: plone/scale/scale.py (hunk 10 above)
Comment (thet) [reply plone/scale/scale.py]: Should be fixed by: https://github.com/plone/plone.scale/pull/39/commits/e77aaa81e2be3f98d25e0c0bd2a33ff4ab086b9d

File: plone/scale/scale.py (hunk 9 above)
Comment (thet) [reply plone/scale/scale.py]: Should be fixed by: https://github.com/plone/plone.scale/pull/39/commits/e77aaa81e2be3f98d25e0c0bd2a33ff4ab086b9d
